iceoryx_hoofs/reporting/include/iox/error_reporting/errors.hpp

```diff
@@ -36,9 +36,21 @@ static constexpr const char* UNKNOWN_ERROR_NAME = "unknown error";
 // By default, there are only error codes and violations.
 // Custom errors can be added but must satisfy the minimal interface.
 
+enum class ViolationErrorCode : uint32_t
+{
+    REQUIRED_CONDITION_VIOLATION,
+    ASSUMPTION_VIOLATION,
+    PRECONDITION_VIOLATION
+};
+
```

Comment:
@elBoberido I looked at #2146 and it is fine (and already merged anyway). One can argue whether more generic arguments including codes would be a good addition in the future, but we can go with this.

Reply:
Looking at how `Expects` and `Ensures` is used in iceoryx I came to the conclusion that it is easily used the wrong™ way and as you also noted it cannot be enforced. Therefore I would propose to just have an `IOX_ASSERT` which is deactivated in release builds. This is something all the developer know and would be a drop in replacement for `assert`. Furthermore, @elfenpiff suggested `IOX_ENFORCE` instead of `IOX_REQUIRE` to emphasis the stronger requirements of always being active.

Reply:
@elBoberido `IOC_ENFORCE` instead of require is fine. We considered it originally but decided against it for reasons I do not remember. 

The same with `IOX_ASSUME` vs `IOX_ASSERT`, but here it was that we considered `IOX_DEBUG_ASSERT` as opposed to `IOX_ASSERT` that would always be active.  This was misleading and so we opted for a new name `IOX_ASSUME`, since it documents the programmers assumption at this point. In the end I fine with either of them as long as there is at least one that can be disabled (usually for performance release mode).

The way of thinking of a code should be: Should this happen in real code unless there is a bug? If yes, then use `IOX_ENFORCE`. Otherwise use `IOX_ASSERT` to document the intention.

As discussed, `Expects` corresponds to `expects`in C++ contracts that were dropped for some reason despite being a good idea (but a bad name IMO). It was meant to check preconditions in some mode (strict,  performance, axiom or something similar). Same with ensures, but for postconditions. The proposal also contained an `assert` of some kind to be used in the same way as `IOX_ASSUME`.

Now we cannot bake it into the language and hence it can be used incorrectly. I would argue it is easy to use correctly and nice to distinguish between preconditions and general assertions and maybe even disable one but not the other (say disable assertions but not preconditions). Also the distinction allows a more fine grained understanding at the handler side of what went wrong. In both cases it should likely be considered a severe error due to a bug, so that might not be that useful. Nevertheless I like to have the option.

However, I am fine with removing it for a first iteration but still do not see the difficulty. If it is an assertion in the sense of above but must hold upon entry in the function, use `IOX_PRECONDITION` (it is not the end of the world if `IOX_ASSERT` is used in its place though). It also stands out in the code and hence one can search for preconditions etc.

I do not think correct use is too much to ask for and the misuse we prevent by dropping `IOX_PRECONDITION` is not a severe case as long as there is still an `IOX_ASSERT`. In the end the end we cannot force the programmer to use any of those. While we can decide to undo this later, I think if it is now decided it will likely stay like this so I wanted to get this "right" from the beginning.

Observe how (IMO) `Expects` and `Ensures` was used incorrectly in the sense that it was not in the spirit of the C++ proposal. This is easily corrected, as the usage sites in where not that many, at least in hoofs. `Ensures` was used completely the wrong way in most cases AFAIR. Currently it serves no purpose, hence why I opted to drop anything specific for  postconditions, an `IOX_ASSERT` is fine here.

Conclusion: 
We can condense it to `IOX_ASSERT` only, but I think this is the wrong approach. Nevertheless we can start like this and reconsider later. Personally I think we should first use it on a larger scale before we make any major or adjustments, unless something clearly does not work. The machinery behind all this makes it easy to define new macros that suit the needs I think. So if the majority would prefer this simpler but IMO less expressive solution we can do this and see how it works out. I am not convinced for reasons above and would keep the distinction between `IOX_PRECONDITION` and `IOX_ASSERT`. All the other renaming is fine.